presto-main/src/main/java/com/facebook/presto/execution/ClusterSizeMonitor.java

```diff
@@ -131,22 +145,62 @@ public void stop()
         return future;
     }
 
+    public synchronized ListenableFuture<?> waitForMinimumCoordinators()
+    {
+        if (currentCoordinatorCount >= minCoordinatorCount) {
+            return immediateFuture(null);
+        }
+
+        SettableFuture<?> future = SettableFuture.create();
+        coordinatorSizeFutures.add(future);
+
+        // if future does not finish in wait period, complete with an exception
+        ScheduledFuture<?> timeoutTask = executor.schedule(
+                () -> {
+                    synchronized (this) {
+                        future.setException(new PrestoException(
+                                GENERIC_INSUFFICIENT_RESOURCES,
+                                format("Insufficient active coordinator nodes. Waited %s for at least %s coordinators, but only %s coordinators are active", executionMaxWait, 2, currentCoordinatorCount)));
+                    }
+                },
+                maxCoordinatorWait.toMillis(),
+                MILLISECONDS);
+
+        // remove future if finished (e.g., canceled, timed out)
+        future.addListener(() -> {
+            timeoutTask.cancel(true);
+            removeFuture(future);
+        }, executor);
+
+        return future;
+    }
+
     private synchronized void removeFuture(SettableFuture<?> future)
     {
-        futures.remove(future);
+        workerSizeFutures.remove(future);
     }
 
     private synchronized void updateAllNodes(AllNodes allNodes)
     {
-        if (includeCoordinator) {
-            currentCount = allNodes.getActiveNodes().size();
-        }
-        else {
-            currentCount = Sets.difference(allNodes.getActiveNodes(), allNodes.getActiveCoordinators()).size();
-        }
+        currentCount = 0;
+        currentCoordinatorCount = 0;
+        allNodes.getActiveNodes().forEach(node -> {
+            if (node.isCoordinator()) {
+                currentCoordinatorCount += 1;
+                if (!includeCoordinator) {
+                    return;
+                }
+            }
+            currentCount += 1;
+        });
         if (currentCount >= executionMinCount) {
```

Comment:
Not sure how important it is, but if the cluster doesn't have enough coordinator and workers, we should always return not enough coordinator error. With current implementation if could be either one (not enough coordinator/not enough worker) of the error that we return and the behavior is not consistent. 
If we can guarantee that somehow, that would be great.

Reply:
We already have this behavior, because one will wait for one or the other via the `waitForMinimumWorkers` or `waitForMinimumCoordinators` methods.